boot/freeldr/bootsect/btrfs.S

```diff
@@ -0,0 +1,790 @@
+    // found chunk map item, checking inclusion with length
+    mov si, CHUNK_MAP_OFFSET
+    lea si, [esi+ecx*8]
+    lea si, [esi+ecx*8]
+    lea si, [esi+ecx*8]   // shift by 24 bytes
+
+    // logical_addr + length
+    mov eax, dword ptr [si]   // low dword of address
+    mov edx, dword ptr [si+4] // high dword of address
+    add eax, dword ptr [si+8] // low dword of length
+    adc edx, dword ptr [si+12] // high dword of length
+    // edx:eax is the end of the chunk
+
+    // (logical_addr + length) - addr_to_find
+    cmp edx, dword ptr [esp+4]
+    ja ConvertAddress_found
+    cmp eax, dword ptr [esp]
+    ja ConvertAddress_found // address is greater than end of the chunk
+    test bl, bl
+    jnz ConvertAddress_notFound
+    ret 8
+
+ConvertAddress_found:
+    // found chunk. Calculating the address
+    // addr_to_find - logical_addr
+    pop eax // low dword of addr_to_find
+    pop edx // high dword of addr_to_find
+    sub eax, dword ptr [si]
+    sbb edx, dword ptr [si+4]
+    // (addr_to_find - logical_addr) + physical_addr
+    add eax, dword ptr [si+16]
+    adc edx, dword ptr [si+20]
+
+    // edx:eax is physical address. Converting to LBA (shifting by 9 bits)
+    shrd eax, edx, 9
+    shr  edx, 9
+    inc bl // clears ZF (bl is 0 or 1 here)
+    ret
+
+ConvertAddress_notFound:
+    // offset is alredy on stack
+    //push dword ptr [esp+4]
+    //push dword ptr [esp]
+    dec sp
+    mov byte ptr [esp], CHUNK_ITEM_KEY
+    data32 push 0
+    push 0
+    push FIRST_CHUNK_TREE_OBJECTID
+
+    mov eax, dword ptr [ChunkRootAddress]
+    mov edx, dword ptr [ChunkRootAddress+4]
+    mov cl, byte ptr [ChunkRootLevel]
+
+    call SearchTree
+    add sp, 9 // setting stack back
+
+    mov ax, si // ES:[SI] - found key pointer
+    call InsertChunk
+
+    mov bl, 0 // indicates second try
+    jmp ConvertAddress_secondTry
+
+
+// Compare key (key1) with key in array identified by base and index (key2)
+// INPUT:
+// - DS:[SI] key1 pointer
+// - ES:[DI] start of the array
+// - AX size of one key in array
+// - CX key index
+// OUTPUT:
+// - DS:[SI] key1 pointer
+// - ES:[DI] key2 pointer
+// - sets flags according to comparison
+CompareKeys:
+    //xchg si, di
+
+    mul cx
+    add di, ax
+    push ds
+    push si
+    push es
+    push di
+
+    // must be replaced for proper flags after cmps
+    push ds
+    push es
+    pop ds
+    pop es
+    xchg si, di
+
+    lea si, [si+4] // key in array
+    lea di, [di+4] // searchable key
+    std
+
+    // comparing objectid
+    cmpsd
+    jne CompareKeys_end
+    cmpsd
+    jne CompareKeys_end
+    // comparing type
+    lea si, [si+12]
+    lea di, [di+12]
+
+    cmpsb
+    jne CompareKeys_end
+    // comparing offset
+    lea si, [si+1+1+4]
+    lea di, [di+1+1+4]
+
+    cmpsd
+    jne CompareKeys_end
+    cmpsd
+CompareKeys_end:
+    cld
+    pop di
+    pop es
+    pop si
+    pop ds
+    ret
+
+
+
+btrfsSignature:
+    .ascii "_BHRfS_M"
+
+//.org 1022
+//    .word HEX(aa55)
```

Comment:
Deprecated stuff?

Reply:
I don't know whether it is needed here or not. We definitely need a signature after the first sector, but what about next ones?